File: ts/earthbucks-lib/src/transaction-signature.ts
export default class TransactionSignature {
  static readonly SIGHASH_ALL = 0x00000001
  static readonly SIGHASH_NONE = 0x00000002
  static readonly SIGHASH_SINGLE = 0x00000003
  static readonly SIGHASH_ANYONECANPAY = 0x00000080

  hashType: number
  sigBuf: Uint8Array

  constructor(hashType: number, sigBuf: Uint8Array) {
    this.hashType = hashType
    this.sigBuf = sigBuf
  }

  toU8Vec(): Uint8Array {
    const hashTypeBuf = Buffer.alloc(1)
    hashTypeBuf.writeUInt8(this.hashType)
    return new Uint8Array(Buffer.concat([hashTypeBuf, this.sigBuf]))
  }

  static fromU8Vec(u8vec: Uint8Array): TransactionSignature {
    const hashType = u8vec[0]
    const sigBuf = u8vec.slice(1)
    return new TransactionSignature(hashType, sigBuf)
  }
}

File: rs/earthbucks_lib/src/transaction_signature.rs
pub struct TransactionSignature {
    pub hash_type: u8,
    pub sig_buf: Vec<u8>,
}

impl TransactionSignature {
    pub const SIGHASH_ALL: u8 = 0x00000001;
    pub const SIGHASH_NONE: u8 = 0x00000002;
    pub const SIGHASH_SINGLE: u8 = 0x00000003;
    pub const SIGHASH_ANYONECANPAY: u8 = 0x00000080;

    pub fn new(hash_type: u8, sig_buf: Vec<u8>) -> Self {
        Self { hash_type, sig_buf }
    }

    pub fn to_u8_vec(&self) -> Vec<u8> {
        let mut result = Vec::new();
        result.push(self.hash_type as u8);
        result.extend(&self.sig_buf);
        result
    }

    pub fn from_u8_vec(data: &[u8]) -> Self {
        let hash_type = data[0];
        let sig_buf = data[1..].to_vec();
        Self { hash_type, sig_buf }
    }
}
